Accessibility/51021 test for basic accessibility for the new and edit (primer) file storage pages

akabiru opened the pull request:
https://community.openproject.org/work_packages/51021

### Accessibility

📖 https://www.openproject.org/docs/development/running-tests/#accessibility-tests

- [x] Use [axe-core-rspec](https://github.com/dequelabs/axe-core-gems/blob/develop/packages/axe-core-rspec/README.md) in automated accessibility tests to provide continuous regression testing against common accessibility issues.
- [x] Use [capybara-accessible-selectors](https://github.com/citizensadvice/capybara_accessible_selectors) in [feature tests](https://www.openproject.org/docs/development/running-tests/#feature-tests) to find UI elements using screen-reader compatible selectors. This ensures the page elements used by feature tests are accessible to assistive technologies.

### Code Maintenance

- [x] Decomposes storage components into testable units, breaking away from convoluted tests at the feature level- rather we have components now fully tested and the feature level focused on user behaviour
- [x] Break up the large `admin_storages_spec` into action specific (index,create,edit) specs

akabiru committed:
[#51021] Test for basic accessibility for the new and edit (primer) file storage pages

https://community.openproject.org/work_packages/51021

akabiru committed:
chore[Op-51021]: Use `click_link_or_button` or `click_on` instead of `click_link`. (convention:Capybara/ClickLinkOrButtonStyle)

akabiru committed:
a11y[Op-51021] update `fill_in` to use form labels in place of IDs

akabiru committed:
a11y[Op-51021]: Heading levels should only increase by one (moderate)

akabiru committed:
Revert "a11y[Op-51021]: Heading levels should only increase by one
(moderate)"

Reverts commit ff6e93d1f25bd384b4a4d5eb94e5478d9d30db0b

Kharonus commented on modules/storages/spec/features/admin_storages_spec.rb:
```diff
@@ -58,9 +58,11 @@
 
+        expect(page).to be_axe_clean.within '#content'
+
         expect(page).to have_test_selector('storage-name', text: complete_storage.name)
```
🟡 should we replace selecting on `test-selector` here, too, whenever it is possible? Point is, that this `data-test-selector` is never visible to any screenreader. So, in theory, we could get false-positives here, when we have a not visible element with that html attribute.

akabiru replied:
~Good point, will replace this with localised `test-selector` 👍🏾~

Edit: We had a chat and the feedback is regarding `have_test_selector` - we settled right now that there doesn't seem to be a better way to make the test more "accessibly declarative" save for scoping `within_test_selector` which is functionally the same thing

aaron-contreras replied:
Would `have_link` not do a better job of what is trying to be accomplished?

akabiru replied:
> Would `have_link` not do a better job of what is trying to be accomplished?

Good idea, it might- but we had a second look and Christophe proposed a really nice rewrite. Will ping once it's up 👍🏾

Kharonus commented on modules/storages/spec/features/admin_storages_spec.rb:
```diff
@@ -58,9 +58,11 @@
+
         expect(page).to have_test_selector('storage-name', text: complete_storage.name)
         expect(page).to have_test_selector('storage-name', text: incomplete_storage.name)
-        expect(page).to have_css("button[aria-label='Add new storage']", text: 'Storage')
```
🟡 should we use `.to have_button` over `.to have_selector :button`?

Kharonus reviewed (commented):
Two basic questions, that could change the overall number of lines to be changed. ^^ Feel free, to share your opinion, as those are the first specs we rewrite with those selectors.

akabiru committed:
chore[Op-51021]: Use `.to have_button` in place of `.to have_selector :button`

aaron-contreras commented on modules/storages/spec/features/admin_storages_spec.rb:
```diff
@@ -107,7 +107,7 @@
 
         # Show Add storage button
-        expect(page).to have_selector :button, 'Storage', aria: { label: 'Add new storage' }
+        expect(page).to have_button 'Storage', aria: { label: 'Add new storage' }
```
Didn't know matchers for node attributes could be specified this way just as in views, neat! 🤌

akabiru committed:
tests[Op#51021]: set timezone preference to prevent failures

In local contexts, there can be failures due to different system
timezones. Explicitly define the timezone preference to mitigate this.

```
expected to find visible css "[data-test-selector=\"storage-health-reason\"]" with text "Unauthorized: Outbound request not authorized since 11/28/2023 01:02 AM" but there were no matches. Also found "Unauthorized: Outbound request not authorized since 11/28/2023 02:02 AM", which matched the selector but not all filters.
```

akabiru committed:
tests[Op#51021]: Reduce time spent in factories via `shared_let`

See: spec/support/shared_let.rb

Before: [TEST PROF INFO] Time spent in factories: 00:00.641 (1.44% of total time)
After: [TEST PROF INFO] Time spent in factories: 00:00.225 (0.85% of total time)

akabiru committed:
chore[Op-51021]: Decompose storage list component, and corresponding feature tests

akabiru committed:
chore[Op#51021]: delegate url_helpers to Rails.application.routes

In view components, the controller context is not always available
specifically when unit testing view components. Similar to the `helpers`
delegation that forwards to action view, introduce similar functionality
for url_helpers.

akabiru committed:
tests[Op-51021] update the correct list item count

akabiru committed:
tests[Op-51021] extract new storage button unit tests from feature

akabiru committed:
tests[Op-51021] update delete spec to use accessible selector

akabiru committed:
tests[Op-51021] Extract health status unit tests to component spec

akabiru committed:
chore[Op#51021]: Break up admin_storages_spec into actions

akabiru committed:
Merge branch 'dev' into accessibility/51021-test-for-basic-accessibility-for-the-new-and-edit-(primer)-file-storage-pages

aaron-contreras commented on modules/storages/spec/features/storages/admin/edit_storage_spec.rb:
```diff
@@ -0,0 +1,318 @@
+                                           text: 'Let OpenProject create folders per project automatically.')
+      end
+
+      aggregate_failures 'General information' do
```
Yes! 👏

akabiru committed:
tests[Op-51021] No need for view component system test helpers

aaron-contreras commented on modules/storages/spec/components/storages/admin/new_storage_button_component_spec.rb:
```diff
@@ -33,11 +33,10 @@
-      expect(page).to have_link 'Nextcloud', href: select_provider_admin_settings_storages_path(provider: 'nextcloud')
-      expect(page).to have_link 'OneDrive/SharePoint', href: select_provider_admin_settings_storages_path(provider: 'one_drive')
-    end
+    expect(page).to have_link 'Nextcloud', href: select_provider_admin_settings_storages_path(provider: 'nextcloud')
+    expect(page).to have_link 'OneDrive/SharePoint', href: select_provider_admin_settings_storages_path(provider: 'one_drive')
```
Unrelated to the change but do we have translations for these?

akabiru replied:
Sure, there are translations- in test we assert against the raw en string for accuracy- i.e. if we just assert against a dynamic value from `I18n` - then if there's a typo in the translations file or the format is incorrect it's relatively difficult to catch

---

See: https://github.com/opf/openproject/blob/dev/modules/storages/app/components/storages/admin/new_storage_button_component.html.erb#L12

Kharonus reviewed (approved):
Love it! This is a huge improvement to test coverage and readability of those tests.
